src/client/signatures/canonical_json/src/ser.rs

```diff
@@ -0,0 +1,486 @@
+use serde::ser::{Serialize};
+use serde_json::ser::{Formatter, CharEscape};
+use std::io::Write;
+use log::{debug, error};
+
+struct JSONFormatter {}
+
+impl Formatter for JSONFormatter {
+    fn write_f64<W: ?Sized>(&mut self, writer: &mut W, value: f64) -> Result<(), std::io::Error> where
+        W: Write,
+    {
+        writer.write(&format_number(value, 10.0_f64.powf(21.0), 10.0_f64.powf(-6.0)).into_bytes())?;
+        Ok(())
+    }
+
+    fn write_char_escape<W: ?Sized>(&mut self, writer: &mut W, char_escape: CharEscape) -> Result<(), std::io::Error> where
+        W: Write, 
+    {
+        match char_escape {
+            CharEscape::Quote => {
+                writer.write(&"\"".to_owned().into_bytes())?; 
```

Comment:
You're converting a `str` into an owned `String`, converting it into bytes, and then taking a reference.  Use `"foo".as_bytes()` instead.

Generally, Rust uses `to_...` to mean a conversion into an owned thing, and `as_...` to mean a zero- (or low-) cost view onto existing data.

Reply:
(Throughout.)

Reply:
I see, yes will make this change